src/payment/bolt12.rs

```diff
@@ -0,0 +1,272 @@
+		if rt_lock.is_none() {
+			return Err(Error::NotRunning);
+		}
+
+		let quantity = None;
+		let amount_msats = None;
+		let mut random_bytes = [0u8; 32];
+		rand::thread_rng().fill_bytes(&mut random_bytes);
+		let payment_id = PaymentId(random_bytes);
+		let retry_strategy = Retry::Timeout(LDK_PAYMENT_RETRY_TIMEOUT);
+		let max_total_routing_fee_msat = None;
+
+		let offer_amount_msat = match offer.amount() {
+			Some(Amount::Bitcoin { amount_msats }) => *amount_msats,
+			Some(_) => {
+				log_error!(self.logger, "Failed to send payment as the provided offer was denominated in an unsupported currency.");
+				return Err(Error::UnsupportedCurrency);
+			},
+			None => amount_msat,
+		};
+
+		if amount_msat < offer_amount_msat {
+			log_error!(
+				self.logger,
+				"Failed to pay as the given amount needs to be at least the offer amount: required {}msat, gave {}msat.", offer_amount_msat, amount_msat);
+			return Err(Error::InvalidAmount);
+		}
+
+		match self.channel_manager.pay_for_offer(
+			&offer,
+			quantity,
+			amount_msats,
+			payer_note,
+			payment_id,
+			retry_strategy,
+			max_total_routing_fee_msat,
+		) {
+			Ok(()) => {
+				let payee_pubkey = offer.signing_pubkey();
+				let offer_amount_msat = offer.amount();
+				log_info!(
+					self.logger,
+					"Initiated sending {:?} to {:?}",
+					offer_amount_msat,
+					payee_pubkey
+				);
+
+				let kind = PaymentKind::Bolt12 { hash: None, preimage: None, secret: None };
+				let payment = PaymentDetails {
+					id: payment_id,
+					kind,
+					amount_msat: Some(amount_msat),
+					direction: PaymentDirection::Outbound,
+					status: PaymentStatus::Pending,
+				};
+				self.payment_store.insert(payment)?;
+
+				Ok(payment_id)
+			},
+			Err(e) => {
+				log_error!(self.logger, "Failed to send payment: {:?}", e);
+				match e {
+					Bolt12SemanticError::DuplicatePaymentId => Err(Error::DuplicatePayment),
+					_ => {
+						let kind = PaymentKind::Bolt12 { hash: None, preimage: None, secret: None };
+						let payment = PaymentDetails {
+							id: payment_id,
+							kind,
+							amount_msat: Some(amount_msat),
+							direction: PaymentDirection::Outbound,
+							status: PaymentStatus::Failed,
+						};
+						self.payment_store.insert(payment)?;
+						Err(Error::PaymentSendingFailed)
+					},
+				}
+			},
+		}
+	}
+
+	/// Returns a payable offer that can be used to request and receive a payment of the amount
+	/// given.
+	pub fn receive(&self, amount_msat: u64, description: &str) -> Result<Offer, Error> {
+		let offer_builder = self.channel_manager.create_offer_builder().map_err(|e| {
+			log_error!(self.logger, "Failed to create offer builder: {:?}", e);
+			Error::OfferCreationFailed
+		})?;
+		let offer = offer_builder
+			.amount_msats(amount_msat)
+			.description(description.to_string())
+			.build()
+			.map_err(|e| {
+				log_error!(self.logger, "Failed to create offer: {:?}", e);
+				Error::OfferCreationFailed
+			})?;
+
+		let id = PaymentId(offer.id().0);
```

Comment:
Not quite sure if I followed the transition plan. IIIUC, someone paying the offer multiple times would result in duplicate payment ids, right?